app/questionnaire_state/answer.py

```diff
@@ -9,9 +9,33 @@ def __init__(self, id):
         # actual user input
         self.input = None
 
-    def update_state(self, user_input):
-        if self.id in user_input.keys():
-            self.input = user_input[self.id]
+    def update_state(self, user_input, schema_item):
+        # Do we have the item or it's containing block?
+        if schema_item.id != self.id:
+            schema_item = schema_item.questionnaire.get_item_by_id(self.id)
+
+        # Get the user input
+        self.input = schema_item.get_user_input(user_input)
+
+        # Mandatory check
+        if schema_item.mandatory and self.input is None:
+            self.errors = []
+            self.errors.append(schema_item.questionnaire.get_error_message('MANDATORY', schema_item.id))
+            self.is_valid = False
+
+        # Try and get the typed value
+        if self.input:
+            try:
+                self.value = schema_item.get_typed_value(user_input)
+                self.is_valid = True
+                self.errors = None
+                self.warnings = None
+            except Exception as e:
```

Comment:
I'm not keen on the use of Exception here, we probably want to define our own exceptions for validation failures